docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,706 @@
+
+```sql
+>set @@foreign_key_checks=0;
+Query OK, 0 rows affected
+>create table t2 (a int key, foreign key fk(a) references t1(id));
+Query OK, 0 rows affected
+>create table t1 (id int key);
+Query OK, 0 rows affected
+>set @@foreign_key_checks=1;
+Query OK, 0 rows affected
+>insert into t2 values (1);
+(1452, 'Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`))')
+```
+
+As you can see, the table `t2` is refer to table `t1`, and when create table `t2`, the table `t1` still not be created, so we can't know the id of table `t1`.
+
+### Create Table with Foreign Key
+
+#### Build TableInfo
+
+When build `TableInfo`, an index for the foreign key columns is created automatically if there is no index covering the foreign key columns. Here is an example:
+
+```sql
+mysql> create table t (id int key, a int, foreign key fk(a) references t(id));
+Query OK, 0 rows affected
+mysql> show create table t\G
+***************************[ 1. row ]***************************
+Table        | t
+Create Table | CREATE TABLE `t` (
+  `id` int NOT NULL,
+  `a` int DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `fk` (`a`),
+  CONSTRAINT `t_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t` (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+1 row in set
+```
+
+As you can see, the index `fk` is created automatically for the foreign key.
+
+#### Validate
+
+Create a table with foreign key, check the following conditions when a DDL job is built and the DDL owner received a DDL job(aka Double-Check):
+
+- whether the user have `REFERENCES` privilege to the foreign key references table.
+- Corresponding columns in the foreign key and the referenced key must have similar data types. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.
+- Supports foreign key references between one column and another within a table. (A column cannot have a foreign key reference to itself.)
+- Indexes on foreign keys and referenced keys are required, because the foreign key check can be fast and do not require a table scan. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.
+- Prefixed indexes on foreign key columns are not supported. Consequently, BLOB and TEXT columns cannot be included in a foreign key because indexes on those columns must always include a prefix length.
+- Does not currently support foreign keys for tables with user-defined partitioning. This includes both reference and child tables.
```

Comment:
What is "user-defined" partitioning?

Reply:
I change to the partition table.
`table with user-defined partitioning` is partition table.
<img width="537" alt="image" src="https://user-images.githubusercontent.com/26020263/207788983-968f1a7c-f9ba-4afa-b6e2-de404506b3a2.png">